src/validation.cpp

```diff
@@ -1283,6 +1297,12 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
         !CheckFeeRate(m_total_vsize, m_total_modified_fees, placeholder_state)) {
         package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package-fee-too-low");
         return PackageMempoolAcceptResult(package_state, {});
+    } else if (args.m_package_feerates &&
+               workspaces.back().m_modified_fees * m_total_vsize < m_total_modified_fees * workspaces.back().m_vsize) {
+        // The package feerate is high enough, but the child's feerate is lower than the package
+        // feerate. This should fail, otherwise we're allowing "parent pay for child."
```

Comment:
Just checking the last tx doesn't guarantee this alone. Maybe mention here that this works because subpackages are being evaluated by themselves as well?

Reply:
I don't think this needs a diamond, even.

For example: mempool is evicting below 10sat/vb, grandparent is A at 3300sat, 100vb (33sat/vb); parent is B spending A, at 700sat, 200vb (3.5sat/vb), child is C, spending B at 2000 sat, 100vb (20sat/vb). If you accept A first, then B alone is still below the eviction threshold, but so is B+C (2700sat, 300vb 9sat/vb), even though C is doing cpfp here.

Reply:
I don't think it's unnecessary - the diamond package would be accepted if we remove this check, the subpackage algorithm doesn't help with this. However, I'm also not sure if  it's sufficient - e.g. what if we'd add another "grandchild" to the child of the diamond package, with a feerate slightly above the previous package feerate (e.g. 5.1 sat/vB). Then this check wouldn't fail anymore, but the descendant feerate of the child (i.e. the combined fees of the child and the grandchild) could be below the minfeerate, so the child/grandchild would get evicted right after submission?

Reply:
The A <- B <- C example is handled without the diamond check, yes. Accept A, reject B, reject B+C. I'm pretty sure the code on master would do this if you removed the child-with-parents requirement.

The case in the unit test isn't handled by validating subpackages, no.
I agree the "no parent pay for child" rule is not enough in some cases, and is too restrictive in others.

For example, let's say the minfeerate is 2sat/vB:

```
A1      A2                 (both 1sat/vB)
  \    /  \
    B1     B2              (both 3.5sat/vB)
      \   /
        C                  (1sat/vB)
```
Ideally we take all except C. 
We reject A1, reject A2, reject A1+A2+B1, accept A2+B2, reject A1+B1+C.
It would be nice if, while considering A1+B1+C, we are again willing to take a subset. Or if we were to try another order or operations, we could reject A1, reject A2, accept A2+B2, accept A1+B1, reject C.
But the logic for these two strategies is essentially "try every possible subset."
For this example, C actually can be any tx we shouldn't accept, e.g. one with an invalid signature.

Perhaps one way to generalize this is that, within an ancestor package, we should be able to accept a subset. But unless we try every subset, it seems there's always a way to rearrange a package or add a tx to make us reject part of it. V3 world has fewer possibilities but still feels a bit intractable.

Maybe we can brainstorm a way to rearrange which checks we do when. Like https://github.com/bitcoin/bitcoin/pull/26711#discussion_r1226067022. Perhaps try everything without feerate checks, eliminate invalid ones, then do feerate analysis. But one transaction could replace transactions in the mempool, then affecting the ancestor set of another transaction in the package...

We can do our best to have the sender-side logic give a good sort. And if a sender is deliberately crafting packages that we'll reject, we address that by taking care not to cache rejections too aggressively. e.g. here we'd want to cache the A1+A2+B1 rejection, but be willing to retry A1+B1.

Reply:
If these retries aren't sufficient to capture CPFPs, they are also unnecessary from the standpoint of getting MVP deployed perhaps. If there is no way to avoid missing things without going full cluster mempool, I'd suggest keeping things very simple to start. I'll let Gloria whiteboard some alternatives before exploring further.

Reply:
> So my https://github.com/bitcoin/bitcoin/pull/26711#discussion_r1226067022 for that would be to have net_processing do retries:

This seems like a good option to consider for computational complexity, but I think could be even slower if we are blocked by UTXO fetching - they are cached when fetched in `PreChecks`, uncached at the end of `ProcessNewPackage`, and then reloaded again when we retry. So just thinking about UTXO fetching, it seems better to do retries while we have them loaded and cached.

I feel like this also requires us making a tradeoff between "good enough result" and potentially a very long queue of things to retry. We would need to put the txns in an orphanage/memory-bounded data structure between retries. We'd want to slow down package relay requests with this peer to avoid getting backed up, be prepared to drop things, etc. In situations where somebody is specially crafting packages, imo we aren't necessarily solving the problem and we'd end up doing more work.